If you need to update this patch, apply it to the old version, then use Git to
rebase it onto the new version.

diff --git a/src/MicroHs/ExpPrint.hs b/src/MicroHs/ExpPrint.hs
index c4328370..565984cd 100644
--- a/src/MicroHs/ExpPrint.hs
+++ b/src/MicroHs/ExpPrint.hs
@@ -31,7 +31,7 @@ removeUnused (ds, emain) = dfs roots M.empty
     dMap = M.fromList ds
     dfs :: [Ident] -> M.Map Exp -> [LDef]
     dfs [] done = M.toList done
-    dfs (i:is) done | Just _ <- M.lookup i done = dfs is done
+    dfs (i:is) done | isJust (M.lookup i done) = dfs is done
                     | otherwise = dfs (freeVars e ++ is) (M.insert i e done)
                                   where e = fromMaybe (error $ "removeUnused: undef " ++ show i) $ M.lookup i dMap
 
@@ -86,7 +86,7 @@ toStringCMdl :: CMdl -> String
 toStringCMdl (ds, emain) =
   let
     def :: (Ident, Exp) -> (String -> String) -> (String -> String)
-    def (i, e) r | Just (_, e', _) <- getForExp e = def (i, e') r
+    def (i, e) r | isJust (getForExp e) = let Just(_, e', _) = getForExp e in def (i, e') r
     def (i, e) r =
       ("A " ++) . toStringP e . ((":" ++ showIdent i ++  " @\n") ++) . r . ("@" ++)
 
diff --git a/src/MicroHs/Expr.hs b/src/MicroHs/Expr.hs
index e12cfd4c..d113864a 100644
--- a/src/MicroHs/Expr.hs
+++ b/src/MicroHs/Expr.hs
@@ -533,7 +533,7 @@ getTupleConstr i =
 getExprTuple :: EType -> Maybe [EType]
 getExprTuple = loop []
   where loop ts (EApp f a) = loop (a:ts) f
-        loop ts (EVar i) | Just n <- getTupleConstr i, length ts == n = Just ts
+        loop ts (EVar i) | isJust (getTupleConstr i) && length ts == fromJust (getTupleConstr i) = Just ts
         loop _ _ = Nothing
 
 -- Create a tuple selector, component i (0 based) of n
@@ -1039,11 +1039,11 @@ instance Pretty Expr where
     ppApp :: PrettyPrec -> [Expr] -> Expr -> Doc
     ppApp prec as (EApp f a) = ppApp prec (a:as) f
     ppApp prec as f | raw = ppApply prec f as
-    ppApp prec as (EVar i) | isOperChar cop, [a, b] <- as = maybeParens (prec > p) $ ppE pl a <+> text op <+> ppE pr b
-                           | isOperChar cop, [a] <- as    = parens $ ppE appPrec a <+> text op
-                           | cop == ',' && length op + 1 == length as
-                                                          = ppE prec (ETuple as)
-                           | op == "[]", length as == 1   = ppE prec (EListish (LList as))
+    ppApp prec as (EVar i) = case as of
+      [a, b] | isOperChar cop -> maybeParens (prec > p) $ ppE pl a <+> text op <+> ppE pr b
+      [a] | isOperChar cop -> parens $ ppE appPrec a <+> text op
+      _ | cop == ',' && length op + 1 == length as -> ppE prec (ETuple as)
+        | op == "[]" && length as == 1 -> ppE prec (EListish (LList as))
       where op = unIdent (unQualIdent i)
             cop = head op
             (pl, p, pr) = lookupPrec op
@@ -1255,5 +1255,5 @@ getImplies _ = Nothing
 
 dropForallContext :: EType -> EType
 dropForallContext (EForall _ _ t) = dropForallContext t
-dropForallContext t | Just (_, t') <- getImplies t = dropForallContext t'
+dropForallContext t | isJust (getImplies t) = let Just (_, t') = getImplies t in dropForallContext t'
                     | otherwise = t
diff --git a/src/MicroHs/FFI.hs b/src/MicroHs/FFI.hs
index cc71597e..3c7de5af 100644
--- a/src/MicroHs/FFI.hs
+++ b/src/MicroHs/FFI.hs
@@ -2,6 +2,7 @@ module MicroHs.FFI(makeFFI) where
 import qualified Prelude(); import MHSPrelude
 import Data.Char
 import Data.List
+import Data.Maybe(fromJust, isJust)
 import MicroHs.Desugar(LDef)
 import MicroHs.Exp
 import MicroHs.Expr
@@ -198,7 +199,8 @@ arity = length . fst . getArrows
 cTypeHsName :: HasCallStack => EType -> String
 cTypeHsName (EApp (EVar ptr) _t) | ptr == identPtr = "Ptr"
                                  | ptr == identFunPtr = "FunPtr"
-cTypeHsName (EVar i) | Just c <- lookup (unIdent i) cHsTypes = c
+cTypeHsName (EVar i) | isJust mc = fromJust mc
+  where mc = lookup (unIdent i) cHsTypes
 cTypeHsName t = errorMessage (getSLoc t) $ "Not a valid C type: " ++ showEType t
 
 cHsTypes :: [(String, String)]
@@ -225,7 +227,8 @@ expTypeHsName _ t = cTypeHsName t
 -- Use to construct 'foreign export ccall' signature.
 cTypeName :: EType -> String
 cTypeName (EApp (EVar ptr) _t) | ptr == identPtr = "void*"
-cTypeName (EVar i) | Just c <- lookup (unIdent i) cTypes = c
+cTypeName (EVar i) | isJust mc = fromJust mc
+  where mc = lookup (unIdent i) cTypes
 cTypeName t = errorMessage (getSLoc t) $ "Not a valid C type: " ++ showEType t
 
 cTypes :: [(String, String)]
@@ -243,7 +246,8 @@ cTypes =
 -- Use to construct 'foreign import javascript' return value wrapper.
 jsTypeNameR :: EType -> String
 jsTypeNameR (EApp (EVar ptr) _) | ptr == identPtr = "PTR"
-jsTypeNameR (EVar i) | Just c <- lookup (unIdent i) jsTypesR = c
+jsTypeNameR (EVar i) | isJust mc = fromJust mc
+  where mc = lookup (unIdent i) jsTypesR
 jsTypeNameR t = errorMessage (getSLoc t) $ "Not a valid Javascript return type: " ++ showEType t
 
 jsTypesR :: [(String, String)]
@@ -256,7 +260,8 @@ jsTypesR =
 -- Use to construct 'foreign import javascript' argument wrapper.
 jsTypeName :: EType -> String
 jsTypeName (EApp (EVar ptr) _) | ptr == identPtr = "Ptr"
-jsTypeName (EVar i) | Just c <- lookup (unIdent i) jsTypes = c
+jsTypeName (EVar i) | isJust mc = fromJust mc
+  where mc = lookup (unIdent i) jsTypes
 jsTypeName t = errorMessage (getSLoc t) $ "Not a valid Javascript argument type: " ++ showEType t
 
 jsTypes :: [(String, String)]
diff --git a/src/MicroHs/Interactive.hs b/src/MicroHs/Interactive.hs
index a1d69107..dc04d7a9 100644
--- a/src/MicroHs/Interactive.hs
+++ b/src/MicroHs/Interactive.hs
@@ -278,9 +278,9 @@ setOptions "+color" = do
   modify $ \ is -> is{ isColor = True }
 setOptions "-color" = do
   modify $ \ is -> is{ isColor = False }
-setOptions s | Just p <- stripPrefix "path=" s =
+setOptions ('p':'a':'t':'h':'=':p) =
   modify $ \ is -> is{ isFlags = (isFlags is){ srcPaths = splitColonPath p } }
-setOptions s | Just p <- stripPrefix "prompt=" s =
+setOptions ('p':'r':'o':'m':'p':'t':'=':p) =
   modify $ \ is -> is{ isPrompt = p }
 setOptions _ =
   putStrLnI "Unknown flag.  Known flags: +s, -s, +c, -c, path=PATH, prompt=STR"
@@ -348,9 +348,10 @@ err e = do
 parseError :: String -> Maybe (FilePath, Int)
 parseError s =
   case words s of
-    "error:" : ('"' : sfile) : "line" : sline : _ |
-      Just file <- stripSuffix "\":" sfile,
-      Just line <- stripSuffix "," sline >>= readMaybe -> Just (file, line)
+    "error:" : ('"' : sfile) : "line" : sline : _ -> do
+      file <- stripSuffix "\":" sfile
+      line <- stripSuffix "," sline >>= readMaybe
+      Just (file, line)
     _ -> Nothing
 
 err' :: String -> I ()
@@ -556,9 +557,11 @@ finds str = do
   let i = mkIdent str
   case stLookup "type" i vs `alt` stLookup "value" i ts of
     Left s -> putStrLnI s
-    Right (Entry (EVar qi) _) | let loc@(SLoc f l _) = slocIdent qi, not (isNoSLoc loc) -> do
-      ed <- getEditor
-      _ <- liftIO $ system $ printf ed l f
-      reload
-      return ()
+    Right (Entry (EVar qi) _) -> case slocIdent qi of
+      loc@(SLoc f l _) | not (isNoSLoc loc) -> do
+        ed <- getEditor
+        _ <- liftIO $ system $ printf ed l f
+        reload
+        return ()
+      _ -> putStrLnI "Unknown location"
     _ -> putStrLnI "Unknown location"
diff --git a/src/MicroHs/Lex.hs b/src/MicroHs/Lex.hs
index 49d70579..1580c3e9 100644
--- a/src/MicroHs/Lex.hs
+++ b/src/MicroHs/Lex.hs
@@ -9,7 +9,7 @@ module MicroHs.Lex(
 import qualified Prelude(); import MHSPrelude hiding(lex)
 import Data.Char
 import Data.List
-import Data.Maybe (fromJust, fromMaybe)
+import Data.Maybe (fromJust, fromMaybe, isJust)
 import MicroHs.Ident
 import Text.ParserComb(TokenMachine(..))
 import Text.PrettyPrint.HughesPJLiteClass(prettyShow)
@@ -144,10 +144,10 @@ lex loc ('(':dcs@(d:cs))
                 par (TSpec _ ')') = p-1
                 par _             = p
 -- Recognize #line 123 "file/name.hs"
-lex loc ('#':xcs) | (SLoc _ _ 1) <- loc, Just cs <- stripPrefix "line " xcs =
+lex loc@(SLoc _ _ 1) ('#':'l':'i':'n':'e':' ':cs) =
   case span (/= '\n') cs of
     (line, rs) -> lex (fromMaybe loc (lineDirective loc line)) rs
-                  | (SLoc _ 1 1) <- loc, take 1 xcs == "!" =
+lex loc@(SLoc _ 1 1) ('#':xcs@('!':cs)) =
   -- It's a shebang (#!), ignore the rest of the line
   skipLine loc xcs
 lex loc ('!':' ':cs) =  -- ! followed by a space is always an operator
@@ -278,8 +278,8 @@ lexLitStr :: SLoc -> SLoc -> ([[Token]] -> String -> [Token]) -> (String -> Mayb
 lexLitStr oloc loc mk end post interp acs = loop loc [] [] acs
   where
         loop :: SLoc -> String -> [[Token]] -> String -> [Token]
-        loop l rs tss cs | Just k <- end cs   = mk (reverse tss) (decodeEscs $ post $ reverse rs) ++ lex (addCol l k) (drop k cs)
-                         | Just (ts, cs', l') <- interp l cs = loop l' (chMark : rs) (ts : tss) cs'
+        loop l rs tss cs | isJust (end cs)    = let Just k = end cs in mk (reverse tss) (decodeEscs $ post $ reverse rs) ++ lex (addCol l k) (drop k cs)
+                         | isJust (interp l cs) = let Just (ts, cs', l') = interp l cs in loop l' (chMark : rs) (ts : tss) cs'
         loop l rs tss ('\\':c:cs) | isSpace c = remGap l rs tss cs
         loop l rs tss ('\\':'^':'\\':cs)      = loop (addCol l 3) ('\\':'^':'\\':rs) tss cs  -- special hack for unescaped \
         loop l rs tss ('\\':cs)               = loop' (addCol l 1) ('\\':rs) tss cs
@@ -322,8 +322,10 @@ decodeEsc ('x':cs) = conv 16 0 cs
 decodeEsc ('o':cs) = conv 8 0 cs
 decodeEsc ('^':c:cs) | '@' <= c && c <= '_' = chr (ord c - ord '@') : decodeEscs cs
 decodeEsc cs@(c:_) | isDigit c = conv 10 0 cs
-decodeEsc (c1:c2:c3:cs) | Just c <- lookup [c1,c2,c3] ctlCodes = c : decodeEscs cs
-decodeEsc (c1:c2:cs) | Just c <- lookup [c1,c2] ctlCodes = c : decodeEscs cs
+decodeEsc (c1:c2:c3:cs) | isJust mc = let Just c = mc in c : decodeEscs cs
+  where mc = lookup [c1,c2,c3] ctlCodes
+decodeEsc (c1:c2:cs) | isJust mc = let Just c = mc in c : decodeEscs cs
+  where mc = lookup [c1,c2] ctlCodes
 decodeEsc (c  :cs) = c : decodeEscs cs
 decodeEsc []       = mhsError "Bad \\ escape"
 
@@ -479,7 +481,7 @@ pragma loc cs =
   in  case map toUpper p of
         "SOURCE" -> TPragma loc p : skip
         -- hsc2hs generates LINE pragmas
-        "LINE" | Just loc' <- lineDirective loc rest -> skipNest loc' 1 ('#':cs)
+        "LINE" | isJust (lineDirective loc rest) -> let Just loc' = lineDirective loc rest in skipNest loc' 1 ('#':cs)
         _ -> skip
 
 -- Parse the arguments of a '#line 123 "file/name.hs"' directive or a LINE pragma.
@@ -501,7 +503,9 @@ lineDirective (SLoc file _ _) cs =
   where
     quoted rs =
       case break (== '"') rs of
-        (_, '"':qs) | (_, '"':nm) <- break (== '"') (reverse qs) -> Just (reverse nm)
+        (_, '"':qs) -> case break (== '"') (reverse qs) of
+          (_, '"':nm) -> Just (reverse nm)
+          _ -> Nothing
         _ -> Nothing
 
 -- | This is the magical layout resolver, straight from the Haskell report.
diff --git a/src/MicroHs/Main.hs b/src/MicroHs/Main.hs
index aff0f432..ada11aba 100644
--- a/src/MicroHs/Main.hs
+++ b/src/MicroHs/Main.hs
@@ -69,14 +69,14 @@ main = do
       let flags' = flags { preload = preload' }
       withArgs rargs $
         case () of
-          _ | Just p <- listPkg flags'        -> mainListPkg flags' p
-          _ | Just p <- buildPkg flags'       -> mainBuildPkg flags' p mdls
-          _ | Just s <- evalArg flags'        -> mainEvalArg flags' s mdls
+          _ | isJust (listPkg flags')         -> let Just p = listPkg flags' in mainListPkg flags' p
+          _ | isJust (buildPkg flags')        -> let Just p = buildPkg flags' in mainBuildPkg flags' p mdls
+          _ | isJust (evalArg flags')         -> let Just s = evalArg flags' in mainEvalArg flags' s mdls
           _ | noCode flags'                   -> mainNoCode flags' mdls
           _ | installPkg flags'               -> mainInstallPackage flags' mdls
           _ | interactive flags'              -> mainInteractive flags' mdls
           _ | null mdls                       -> mainCompileC flags' [] ""
-          _ | [s] <- mdls                     -> mainCompile flags' (mkIdentSLoc (SLoc "command-line" 0 0) s)
+          _ | length mdls == 1                -> mainCompile flags' (mkIdentSLoc (SLoc "command-line" 0 0) $ head mdls)
           _                                   -> mhsError usage
 
 mHSPKG :: String
@@ -161,25 +161,25 @@ decodeArgs f mdls (arg:args) =
     "-z"        -> decodeArgs f{compress = True} mdls args
     "-b64"      -> decodeArgs f{base64 = True} mdls args
     "-Q"        -> decodeArgs f{installPkg = True} mdls args
-    "-o" | s : args' <- args
-                -> decodeArgs f{output = s} mdls args'
-    "-optc" | s : args' <- args
-                -> decodeArgs f{cArgs = cArgs f ++ [s]} mdls args'
-    "-optl" | s : args' <- args
-                -> decodeArgs f{lArgs = lArgs f ++ [s]} mdls args'
-    "-optF" | s : args' <- args
-                -> decodeArgs f{fArgs = fArgs f ++ [s]} mdls args'
-    "-pgmF" | s : args' <- args
-                -> decodeArgs f{fPgm = Just s} mdls args'
-    "-interactive-print" | s : args' <- args
-                -> decodeArgs f{iPrint = Just s} mdls args'
+    "-o" | not (null args)
+                -> let s : args' = args in decodeArgs f{output = s} mdls args'
+    "-optc" | not (null args)
+                -> let s : args' = args in decodeArgs f{cArgs = cArgs f ++ [s]} mdls args'
+    "-optl" | not (null args)
+                -> let s : args' = args in decodeArgs f{lArgs = lArgs f ++ [s]} mdls args'
+    "-optF" | not (null args)
+                -> let s : args' = args in decodeArgs f{fArgs = fArgs f ++ [s]} mdls args'
+    "-pgmF" | not (null args)
+                -> let s : args' = args in decodeArgs f{fPgm = Just s} mdls args'
+    "-interactive-print" | not (null args)
+                -> let s : args' = args in decodeArgs f{iPrint = Just s} mdls args'
     "-F"        -> decodeArgs f{doF = True} mdls args
     "--stdin"   -> decodeArgs f{useStdin = True} mdls args
     "--interactive"   -> decodeArgs f{interactive = True} mdls args
-    "--embed-ffis" | s : args' <- args
-                -> decodeArgs f{embedFFIs = embedFFIs f ++ splitColonPath s} mdls args'
-    "--embed-packages" | s : args' <- args, let ps = splitColonPath s
-                -> decodeArgs f{embedPkgs = embedPkgs f ++ ps, embedFFIs = embedFFIs f ++ ps} mdls args'
+    "--embed-ffis" | not (null args)
+                -> let s : args' = args in decodeArgs f{embedFFIs = embedFFIs f ++ splitColonPath s} mdls args'
+    "--embed-packages" | not (null args)
+                -> let s : args' = args; ps = splitColonPath s in decodeArgs f{embedPkgs = embedPkgs f ++ ps, embedFFIs = embedFFIs f ++ ps} mdls args'
     "-fno-code" -> decodeArgs f{noCode = True} mdls args
     '-':'i':[]  -> decodeArgs f{srcPaths = []} mdls args
     '-':'i':s   -> decodeArgs f{srcPaths = srcPaths f ++ splitColonPath s} mdls args
@@ -194,7 +194,7 @@ decodeArgs f mdls (arg:args) =
     '-':'p':s   -> decodeArgs f{preload = preload f ++ [s]} mdls args
     '-':'E':s   -> decodeArgs f{editor = Just s} mdls args
     '-':'e':s   -> decodeArgs f{evalArg = Just s} mdls args
-    _ | Just r  <- stripPrefix "-ddump-" arg, Just d <- lookup r dumpFlagTable ->
+    '-':'d':'d':'u':'m':'p':'-':r | isJust (lookup r dumpFlagTable) -> let Just d = lookup r dumpFlagTable in
                    decodeArgs f{dumpFlags = d : dumpFlags f} mdls args
 
     '-':_       -> mhsError $ "Unknown flag: " ++ arg ++ "\n" ++ usage
diff --git a/src/MicroHs/TypeCheck.hs b/src/MicroHs/TypeCheck.hs
index 62ddda14..df002e5d 100644
--- a/src/MicroHs/TypeCheck.hs
+++ b/src/MicroHs/TypeCheck.hs
@@ -834,7 +834,7 @@ tInst :: HasCallStack => Expr -> EType -> T (Expr, EType)
 tInst ae (EForall _ vks t) = do
   t' <- tInstForall vks t
   tInst ae t'
-tInst ae at | Just (ctx, t) <- getImplies at = do
+tInst ae at | isJust (getImplies at) = let Just (ctx, t) = getImplies at in do
   --tcTrace $ "tInst: addConstraint: " ++ show ae ++ ", " ++ show d ++ " :: " ++ show ctx
 {-
   if eqExpr ae eCannotHappen then
@@ -1342,7 +1342,7 @@ expandClass d = return [d]
 -- Ignoring initial quantifiers and context, how many arrows does the type have?
 countArrows :: EType -> Int
 countArrows (EForall _ _ t) = countArrows t
-countArrows t | Just (_, t') <- getImplies t = countArrows t'
+countArrows t | isJust (getImplies t) = let Just (_, t') = getImplies t in countArrows t'
               | otherwise = length . fst . getArrows $ t
 
 simpleEqn :: Expr -> [Eqn]
@@ -1819,7 +1819,7 @@ tInferExpr = tInfer tcExpr
 
 tCheckExpr :: HasCallStack =>
               EType -> Expr -> T Expr
-tCheckExpr t e | Just (ctx, t') <- getImplies t = do
+tCheckExpr t e | isJust (getImplies t) = let Just (ctx, t') = getImplies t in do
 --  tcTrace $ "tCheckExpr: " ++ show (e, ctx, t')
   xt <- expandSyn t
   unless (eqEType t xt) undefined
@@ -2177,7 +2177,7 @@ needDsCase ex lbls ae =
       vt <- gets valueTable
       case stLookupGlbUnqMany lbl vt of
         -- check the return type of an unambiguous label
-        Just [Entry _ t] | (t':_, _) <- getArrows (dropForallContext t) -> needT t'
+        Just [Entry _ t] | not $ null $ fst $ getArrows (dropForallContext t) -> needT $ head $ fst $ getArrows $ dropForallContext t
         _ -> return Nothing
 
 -- Do a record update using case.
@@ -2236,8 +2236,8 @@ tcExprAp mt ae args = do
                  EUVar r -> fmap (fromMaybe t) (getUVar r)
                  _ -> return t
 --             tcTrace $ "exExprAp: EVar " ++ showIdent i ++ " :: " ++ showExpr t ++ " = " ++ showExpr t' ++ " mt=" ++ show mt
-             case fn of
-               EVar ii | ii == mkIdent "Data.Function.$", f:as <- args -> tcExprAp mt f as
+             case (fn, args) of
+               (EVar ii, f:as) | ii == mkIdent "Data.Function.$" -> tcExprAp mt f as
                _ -> tcExprApFn mt fn t' args
     EQVar f t ->  -- already resolved
       tcExprApFn mt f t args
@@ -2268,17 +2268,17 @@ tcExprApFn mt fn atfn aargs = do
   let -- loop _ats aas ft | trace ("loop: " ++ show (aas, ft)) False = undefined
       loop ats [] ft = final (reverse ats) ft
       loop ats aas@(a:as) aft = do
-        case nextArg aft of
-          AReqd (IdKind i k) ft -> useType i k a ft
-          AForall _ (IdKind i k:iks) ft | ETypeArg t <- a -> do
+        case (nextArg aft, a) of
+          (AReqd (IdKind i k) ft, _) -> useType i k a ft
+          (AForall _ (IdKind i k:iks) ft, ETypeArg t) -> do
             -- traceM ("AForall " ++ show (i, t))
             useType i k t (EForall QExpl iks ft)
-          AForall _ iks ft -> do
+          (AForall _ iks ft, _) -> do
             ft' <- tInstForall iks ft
             loop ats aas ft'
-          AConstaint ctx ft ->
+          (AConstaint ctx ft, _) ->
             loop (ArgCtx ctx : ats) aas ft
-          ARet aft' -> do
+          (ARet aft', _) -> do
             (at, rt) <- unArrow loc aft'
             --traceM ("ARet " ++ show (at, rt))
             loop (ArgExpr a at : ats) as rt
@@ -2521,7 +2521,7 @@ tcOper ae aies = do
   fixs <- gets fixTable
   let
     opfix :: (Ident, Expr) -> T ((Expr, Fixity), Expr)
-    opfix (i, e) | Just 2 <- getTupleConstr i =  -- tcLSect/tcRSect can introduce ',' as an operator
+    opfix (i, e) | Just 2 == getTupleConstr i =  -- tcLSect/tcRSect can introduce ',' as an operator
       return ((EVar i, (AssocNone, -1)), e)
     opfix (i, e) = do
       (ei, _) <- tLookupV i
@@ -2602,7 +2602,7 @@ nextArg :: EType -> Arg
 nextArg (EForall _ []  t)                  = nextArg t
 nextArg (EForall QReqd (ik:iks) t)         = AReqd ik (EForall QReqd iks t)
 nextArg (EForall q     iks      t)         = AForall q iks t
-nextArg t | Just (ctx, t') <- getImplies t = AConstaint ctx t'
+nextArg t | isJust (getImplies t) = let Just (ctx, t') = getImplies t in AConstaint ctx t'
           | otherwise                      = ARet t
 
 tcExprLam :: HasCallStack => Expected -> SLoc -> [Eqn] -> T Expr
@@ -2618,7 +2618,7 @@ tcEqns' top at eqns =
   case at of
     EForall QExpl iks t -> withExtTyps iks $ tcEqns' top t eqns
     EForall QImpl   _ t ->                   tcEqns' top t eqns
-    _ | Just (ctx, t') <- getImplies at -> do
+    _ | isJust (getImplies at) -> let Just (ctx, t') = getImplies at in do
       let loc = getSLoc eqns
       d <- newADictIdent loc
       f <- newIdent loc "fcnD"
@@ -3173,7 +3173,7 @@ skolemise (EForall _ tvs ty) = do -- Rule PRPOLY
   (sks1, ty') <- shallowSkolemise tvs ty
   (sks2, ty'') <- skolemise ty'
   return (sks1 ++ sks2, ty'')
-skolemise t@(EApp _ _) | Just (arg_ty, res_ty) <- getArrow t = do
+skolemise t@(EApp _ _) | isJust (getArrow t) = let Just (arg_ty, res_ty) = getArrow t in do
   (sks, res_ty') <- skolemise res_ty
   return (sks, arg_ty `tArrow` res_ty')
 skolemise (EApp f a) = do
@@ -3242,13 +3242,13 @@ subsCheckRho loc exp1 (EForall _ vs1 t1) (EForall _ vs2 t2) | length vs1 == leng
 subsCheckRho loc exp1 sigma1@EForall{} rho2 = do -- Rule SPEC
   (exp1', rho1) <- tInst exp1 sigma1
   subsCheckRho loc exp1' rho1 rho2
-subsCheckRho loc exp1 arho1 rho2 | Just _ <- getImplies arho1 = do
+subsCheckRho loc exp1 arho1 rho2 | isJust (getImplies arho1) = do
   (exp1', rho1) <- tInst exp1 arho1
   subsCheckRho loc exp1' rho1 rho2
-subsCheckRho loc exp1 rho1 rho2 | Just (a2, r2) <- getArrow rho2 = do -- Rule FUN
+subsCheckRho loc exp1 rho1 rho2 | isJust (getArrow rho2) = let Just (a2, r2) = getArrow rho2 in do -- Rule FUN
   (a1, r1) <- unArrow loc rho1
   subsCheckFun loc exp1 a1 r1 a2 r2
-subsCheckRho loc exp1 rho1 rho2 | Just (a1, r1) <- getArrow rho1 = do -- Rule FUN
+subsCheckRho loc exp1 rho1 rho2 | isJust (getArrow rho1) = let Just (a1, r1) = getArrow rho1 in do -- Rule FUN
   (a2,r2) <- unArrow loc rho2
   subsCheckFun loc exp1 a1 r1 a2 r2
 subsCheckRho loc exp1 tau1 tau2 = do  -- Rule MONO
@@ -3430,10 +3430,12 @@ canonPatSynType at = do
       pure $ mkTyp [] emptyCtx [] emptyCtx ty
 
 splitPatSynType :: EType -> ([IdKind], EConstraint, [IdKind], EConstraint, EType)
-splitPatSynType (EForall _ vks1 t0)
-  | Just  (ctx1, EForall _ vks2 t1) <- getImplies t0
-  , Just  (ctx2, ty) <- getImplies t1
-  = (vks1, ctx1, vks2, ctx2, ty)
+splitPatSynType (EForall _ vks1 t0) | isJust x = fromJust x
+  where
+    x = do
+      (ctx1, EForall _ vks2 t1) <- getImplies t0
+      (ctx2, ty) <- getImplies t1
+      Just (vks1, ctx1, vks2, ctx2, ty)
 splitPatSynType t = impossiblePP t
 
 -----
@@ -3529,7 +3531,7 @@ defaultOneTyVar tv = do
 --  traceM $ "defaultOneTyVar: cvs = " ++ show cvs
   dvs <- getSuperClasses cvs                            -- add superclasses
 --  traceM $ "defaultOneTyVar: dvs = " ++ show dvs
-  let oneCls c | Just ts <- M.lookup c (defaults old) =
+  let oneCls c | isJust (M.lookup c $ defaults old) = let Just ts = M.lookup c (defaults old) in
         take 1 $ filter (\ t -> all (\ cc -> soluble cc t) cvs) ts
                | otherwise = []
       soluble c t = fst $ flip tcRun old $ do
@@ -3634,7 +3636,7 @@ solvers =
 -- Examine each goal, either solve it (possibly producing new goals) or let it remain unsolved.
 solveMany :: [Goal] -> [UGoal] -> [(EType, Soln)] -> [Improve] -> T ([UGoal], [Soln], [Improve])
 solveMany [] uns sol imp = return (uns, map snd sol, imp)
-solveMany ((di, ct) : cnss) uns sol imp | Just (_, (dd, _)) <- find (eqEType ct . fst) sol =
+solveMany ((di, ct) : cnss) uns sol imp | isJust (find (eqEType ct . fst) sol) = let Just (_, (dd, _)) = find (eqEType ct . fst) sol in
   solveMany cnss uns ((ct, (di, EVar dd)) : sol) imp
 -- Need to handle ct of the form C => T, and forall a . T
 solveMany (cns@(di, ct) : cnss) uns sol imp = do
@@ -3970,7 +3972,7 @@ solveEq eqs t1 t2 | normTypeEq eqs t1 `eqEType` normTypeEq eqs t2 = Just []
 -- XXX This guaranteed by how it's called, but I'm not sure it always works properly.
 addTypeEq :: EType -> EType -> TypeEqTable -> TypeEqTable
 addTypeEq t1 t2 aeqs =
-  let deref (EVar i) | Just t <- lookup i aeqs = t
+  let deref (EVar i) | isJust (lookup i aeqs) = fromJust $ lookup i aeqs
       deref (ESign t _) = t
       deref t = t
       t1' = deref t1
@@ -4071,7 +4073,7 @@ standaloneDeriving str narg act = do
 --  traceM ("standaloneDeriving 1 " ++ show (_vks, _ctx, cc))
   (cls, ts, tname) <-
     case getAppM cc of
-      Just (c, ts@(_:_)) | Just (n, _) <- getAppM (last ts) -> return (c, init ts, n)
+      Just (c, ts@(_:_)) | isJust (getAppM $ last ts) -> let Just (n, _) = getAppM (last ts) in return (c, init ts, n)
       _ -> tcError (getSLoc act) "malformed standalone deriving"
 --  traceM ("standaloneDeriving 2 " ++ show (act, cls, tname))
   dtable <- gets dataTable
diff --git a/src/Text/ParserComb.hs b/src/Text/ParserComb.hs
index 6efd5a3f..8ab72759 100644
--- a/src/Text/ParserComb.hs
+++ b/src/Text/ParserComb.hs
@@ -127,7 +127,9 @@ satisfy msg f = P $ \ acs ->
 satisfyM :: forall tm t a . TokenMachine tm t => String -> (t -> Maybe a) -> Prsr tm t a
 satisfyM msg f = P $ \ acs ->
   case tmNextToken acs of
-    (c, cs) | Just a <- f c -> Success a cs noFail
+    (c, cs) -> case f c of
+        Just a -> Success a cs noFail
+        _ -> Failure (LastFail (tmLeft acs) (firstToken acs) [msg])
     _ -> Failure (LastFail (tmLeft acs) (firstToken acs) [msg])
 
 infixl 9 <?>
